hyperhtml.js

```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
     };
   }
 
-  // update attributes node
-  //
-  // render`<a href="${url}" onclick="${click}">${name}</a>`;
-  //
-  // Note: attributes with a special meaning like DOM Level 0
-  //       listeners or accessors properties are directly set
-  function setAttribute(node, attribute) {
+  // `<div class="${'attr'}"></div>`
+  function setAttribute(attribute, removeAttributes) {
     var
       name = attribute.name,
+      node = attribute.ownerElement,
       isEvent = name.slice(0, 2) === 'on',
       isSpecial = name in node && !SHOULD_USE_ATTRIBUTE.test(name),
       oldValue
     ;
-    if (isSpecial) node.removeAttribute(name);
+    if (isSpecial) removeAttributes.push(node, name);
     return isSpecial ?
       function specialAttr(newValue) {
         if (oldValue !== newValue) {
+          if (isEvent && oldValue && ('handleEvent' in oldValue)) {
+            handleEvent(node, 'remove', name, oldValue);
+          }
           oldValue = newValue;
-          if (isEvent) {
-            node[name] = 'handleEvent' in newValue ?
-              newValue.handleEvent.bind(newValue) :
-              newValue;
+          if (isEvent && newValue && ('handleEvent' in newValue)) {
+            handleEvent(node, 'add', name, newValue);
           } else {
             node[name] = newValue;
           }
         }
       } :
       function attr(newValue) {
         if (oldValue !== newValue) {
-          attribute.value = (oldValue = newValue);
+          oldValue = newValue;
+          attribute.value = newValue;
         }
       };
   }
 
-  // update the "emptiness"
-  // this function is used when template literals
-  // have sneaky html/fragment capable
-  // updates in the wild (no spaces around)
-  //
-  // render`
-  //  <p>Content before</p>${
-  //  'any content in between'
-  //  }<p>Content after</p>
-  // `;
-  //
-  // Note: this is the most expensive
-  //       update of them all.
-  function setVirtualContent(node) {
-    var
-      fragment = node.ownerDocument.createDocumentFragment(),
-      childNodes = []
-    ;
+  // `<div> ${'text'} </div>`
+  function setTextContent(node) {
+    var oldValue;
+    return function text(newValue) {
+      if (oldValue !== newValue) {
+        oldValue = newValue;
+        node.textContent = newValue;
+      }
+    };
+  }
+
+  // `<li>a</li>${'virtual'}<li>c</li>`
+  function setVirtualContent(node, childNodes) {
```

Comment:
Maybe add a bit more of clarification of what you mean by "virtual", as it's an overloaded word.